numpyro/infer/elbo.py

```diff
@@ -454,76 +456,40 @@ def dfs(site, visited):
     return top_sorted if reverse else list(reversed(top_sorted))
 
 
-def _compute_downstream_costs(model_trace, guide_trace, non_reparam_nodes):
-    model_successors = _identify_dense_edges(model_trace)
-    guide_successors = _identify_dense_edges(guide_trace)
-    # recursively compute downstream cost nodes for all sample sites in model and guide
-    # (even though ultimately just need for non-reparameterizable sample sites)
-    # 1. downstream costs used for rao-blackwellization
-    # 2. model observe sites (as well as terms that arise from the model and guide having different
-    # dependency structures) are taken care of via 'children_in_model' below
-    topo_sort_guide_nodes = _topological_sort(guide_successors, reverse=True)
-    topo_sort_guide_nodes = [
-        x for x in topo_sort_guide_nodes if guide_trace[x]["type"] == "sample"
-    ]
-    ordered_guide_nodes_dict = {n: i for i, n in enumerate(topo_sort_guide_nodes)}
-
-    downstream_guide_cost_nodes = {}
-    downstream_costs = {}
-    stacks = _get_plate_stacks(model_trace)
-
-    for node in topo_sort_guide_nodes:
-        downstream_costs[node] = MultiFrameTensor(
-            (
-                stacks[node],
-                model_trace[node]["log_prob"] - guide_trace[node]["log_prob"],
+class track_nonreparam(Messenger):
+    def postprocess_message(self, msg):
+        # track non-reparameterizable sample sites
+        if (
+            msg["type"] == "sample"
+            and not msg["is_observed"]
+            and not getattr(msg["fn"], "has_rsample", False)
+        ):
+            new_provenance = frozenset({msg["name"]})
+            old_provenance = msg["value"].aval.named_shape.get(
+                "_provenance", frozenset()
             )
```

Comment:
a -> b -> c

this function - prob(c): (a, b, c)

render_model - prob(c): (b, c)
If we know prob(b): (a, b) => prob(c): (a, b, c)